### PR TITLE
libbsd: support Conan v2

### PR DESCRIPTION
**Pull request by Tobulus**
Specify library name and version:  **libbsd/0.10.0**

Updating libbsd for Conan x2

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libbsd/all/conanfile.py [hunk 1]
```diff
@@ -32,51 +36,70 @@ def build_requirements(self):
...
+            tc.extra_cflags.append("-Wno-error=implicit-function-declaration")
+        tc.generate()
+        deps = AutotoolsDeps(self)
+        deps.generate()
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
     
     def validate(self):
-        if not tools.is_apple_os(self.settings.os) and self.settings.os != "Linux":
```
Comment (Tobulus) [inline recipes/libbsd/all/conanfile.py:42]: fails on MacOS due to this: https://gitlab.freedesktop.org/libbsd/libbsd/-/issues/1

don't know why this was building before?

File: recipes/libbsd/all/conanfile.py [hunk 2]
```diff
@@ -32,51 +36,70 @@ def build_requirements(self):
...
+            del self.settings.compiler.cppstd
+        except ConanException:
+            pass
+    
+    def generate(self):
+        if not cross_building(self):
+            env = VirtualRunEnv(self)
+            env.generate(scope="build")
+        tc = AutotoolsToolchain(self)
+        if is_apple_os(self):
```
Comment (Tobulus) [inline recipes/libbsd/all/conanfile.py:59]: redundant if MacOS is not supported

File: recipes/libbsd/all/conanfile.py (hunk 1 above)
Comment (Tobulus) [reply recipes/libbsd/all/conanfile.py:42]: see this build run: https://github.com/conan-io/conan-center-index/pull/16893#issuecomment-1509174805

File: recipes/libbsd/all/conanfile.py [hunk 3]
```diff
@@ -32,51 +36,70 @@ def build_requirements(self):
...
+
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
-        del self.settings.compiler.libcxx
-        del self.settings.compiler.cppstd
+        try:
+            del self.settings.compiler.libcxx
+        except ConanException:
+            pass
```
Comment (prince-chrismc) [inline recipes/libbsd/all/conanfile.py:48]: This can be relaced by `rm_safe`

File: recipes/libbsd/all/conanfile.py [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
 import os
 
+from conan import ConanFile
+from conan.errors import ConanException, ConanInvalidConfiguration
+from conan.tools.gnu import Autotools, AutotoolsToolchain, AutotoolsDeps
+from conan.tools.apple import is_apple_os
+from conan.tools.env import VirtualRunEnv
+from conan.tools.build import cross_building
+from conan.tools.files import apply_conandata_patches, export_conandata_patches, copy, get, rmdir
+from conan.tools.cmake import cmake_layout
```
Comment (prince-chrismc) [inline recipes/libbsd/all/conanfile.py:11]: There should be a min conan version of 1.53.0 here

File: recipes/libbsd/all/test_v1_package/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package C)
+
+include("${CMAKE_BINARY_DIR}/conanbuildinfo.cmake")
+conan_basic_setup(TARGETS)
+
+find_package(libbsd REQUIRED)
```
Comment (prince-chrismc) [inline recipes/libbsd/all/test_v1_package/CMakeLists.txt:7]: You could drop the duplicated test_package.c if you update this to the template

https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/test_v1_package/CMakeLists.txt

File: recipes/libbsd/all/conanfile.py [hunk 6]
```diff
@@ -21,7 +26,6 @@ class LibBsdConan(ConanFile):
         "shared": False,
         "fPIC": True,
     }
-    exports_sources = "patches/**"
 
     _autotools = None
```
Comment (prince-chrismc) [inline recipes/libbsd/all/conanfile.py:30]: This is not needed with the new helpers and the generate method

File: recipes/libbsd/all/conanfile.py [hunk 7]
```diff
@@ -32,51 +36,70 @@ def build_requirements(self):
     def _source_subfolder(self):
         return "source_subfolder"
```
Comment (prince-chrismc) [inline recipes/libbsd/all/conanfile.py:37]: This should be replaced with a basic_layout

File: recipes/libbsd/all/conanfile.py [hunk 8]
```diff
@@ -32,51 +36,70 @@ def build_requirements(self):
...
+            env.generate(scope="build")
+        tc = AutotoolsToolchain(self)
+        if is_apple_os(self):
+            tc.extra_cflags.append("-Wno-error=implicit-function-declaration")
+        tc.generate()
+        deps = AutotoolsDeps(self)
+        deps.generate()
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
```
Comment (prince-chrismc) [inline recipes/libbsd/all/conanfile.py:66]: This is not a Cmake project

```suggestion
        basic_layout(self, src_folder="src")
```

File: recipes/libbsd/all/conanfile.py [hunk 9]
```diff
@@ -32,51 +36,70 @@ def build_requirements(self):
...
     
     def validate(self):
-        if not tools.is_apple_os(self.settings.os) and self.settings.os != "Linux":
-            raise ConanInvalidConfiguration("libbsd is only available for GNU-like operating systems (e.g. Linux)")
+        if self.settings.os != "Linux":
+            raise ConanInvalidConfiguration("libbsd is only available for Linux currently")
```
Comment (prince-chrismc) [inline recipes/libbsd/all/conanfile.py:70]:
```suggestion
            raise ConanInvalidConfiguration(f"{self.ref} is only available for Linux currently")
```

File: recipes/libbsd/all/conanfile.py (hunk 2 above)
Comment (prince-chrismc) [reply recipes/libbsd/all/conanfile.py:59]: 😕 https://github.com/conan-io/conan-center-index/blob/0cfba12866ccf53adfe6eb6fcc228f397801aba6/recipes/libbsd/all/conandata.yml#L7 

it looks like support is being patched in?

**Review by prince-chrismc — COMMENTED** [review]
Handleful of comments, this recipes needs love so thanks for helping update it

File: recipes/libbsd/all/conanfile.py (hunk 2 above)
Comment (Tobulus) [reply recipes/libbsd/all/conanfile.py:52]: your are right, I missed that. Let me take a look...

File: recipes/libbsd/all/conanfile.py (hunk 6 above)
Comment (Tobulus) [reply recipes/libbsd/all/conanfile.py:32]: As this is used in build() and package(): do you have an example how to handle this?

File: recipes/libbsd/all/conanfile.py [hunk 10]
```diff
@@ -21,62 +28,63 @@ class LibBsdConan(ConanFile):
...
+        self.settings.rm_safe("compiler.libcxx")
+        self.settings.rm_safe("compiler.cppstd")
+    
+    def generate(self):
+        if not cross_building(self):
+            env = VirtualRunEnv(self)
+            env.generate(scope="build")
+        tc = AutotoolsToolchain(self)
+        if is_apple_os(self):
+            tc.extra_cflags.append("-Wno-error=implicit-function-declaration")
```
Comment (Tobulus) [inline recipes/libbsd/all/conanfile.py:51]: this cflag is not used correctly. @prince-chrismc any idea what I'm doing wrong?

Comment (Tobulus) [reply recipes/libbsd/all/conanfile.py:50]: It is somehow related to the autoreconf() call

Comment (Tobulus) [reply recipes/libbsd/all/conanfile.py:48]: Done

File: recipes/libbsd/all/conanfile.py (hunk 6 above)
Comment (Tobulus) [reply recipes/libbsd/all/conanfile.py]: Done

**Review by prince-chrismc — DISMISSED** [review]
Lots of good work!

File: recipes/libbsd/all/conanfile.py [hunk 11]
```diff
@@ -21,72 +28,66 @@ class LibBsdConan(ConanFile):
...
-    def _source_subfolder(self):
-        return "source_subfolder"
+        self.tool_requires("libtool/2.4.7")
 
+    def export_sources(self):
+        export_conandata_patches(self)
+    
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
```
Comment (prince-chrismc) [inline recipes/libbsd/all/conanfile.py:40]:
```suggestion
           self.options.rm_safe("fPIC")
```

**Review by prince-chrismc — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

**Pull request merged**